### PR TITLE
Top level URI Variables

### LINKED ISSUE #1357: UriVariables can't be used for RootForms
Hi.

I just found out that [uriVariables](https://w3c.github.io/wot-thing-description/#form-uriVariables) can currently only be defined in [InteractionAffordances](https://w3c.github.io/wot-thing-description/#interactionaffordance).
That rules out using them for "RootForms" - which also is defined in the JsonSchema: 
* `property_element`: https://github.com/w3c/wot-thing-description/blob/9db5e96f5d8654e54a2d6884b90485251b9768da/validation/td-json-schema-validation.json#L536-L541
* `action_element`: https://github.com/w3c/wot-thing-description/blob/9db5e96f5d8654e54a2d6884b90485251b9768da/validation/td-json-schema-validation.json#L652-L657
* `event_element`: https://github.com/w3c/wot-thing-description/blob/9db5e96f5d8654e54a2d6884b90485251b9768da/validation/td-json-schema-validation.json#L701-L706

I wonder if this is an explicit decision or if this just was not yet thought about.
Using `uriVariables` in root forms might become really useful, e.g. in order to define which `fields` to select for the `readmultipleproperties` OP, e.g.:
```json
{
    "@context": [ "https://www.w3.org/2022/wot/td/v1.1" ],
    "security": "basic_sc",
    "id": "urn:de.jaeckles.test:the-thing-1",
    "base": "http://localhost:8080/api/2/things/de.jaeckles.test:the-thing-1",
    "securityDefinitions": {
        "basic_sc": {
            "in": "header",
            "scheme": "basic"
        }
    },
    "uriVariables": {
        "fields": {
            "type": "string",
            "title": "Comma separated list of properties to select."
        }
    },
    "forms": [
        {
            "op": "readallproperties",
            "href": "/attributes",
            "htv:methodName": "GET"
        },
        {
            "op": "writeallproperties",
            "href": "/attributes",
            "htv:methodName": "PUT"
        },
        {
            "op": "readmultipleproperties",
            "href": "/attributes{?fields}",
            "htv:methodName": "GET"
        },
        {
            "op": "writemultipleproperties",
            "href": "/attributes",
            "htv:methodName": "PATCH",
            "contentType": "application/merge-patch+json"
        },
        {
            "op": [
                "observeallproperties",
                "unobserveallproperties"
            ],
            "href": "/attributes",
            "htv:methodName": "GET",
            "subprotocol": "sse",
            "contentType": "text/event-stream"
        }
    ],
...
```

Which could be usable as:
```
GET http://localhost:8080/api/2/things/de.jaeckles.test:the-thing-1/attributes?fields=location,manufacturer
```

### PR DESCRIPTION
**Pull request by egekorkan**
fixes #1357 

I have added uriVariables to Thing level in the information model and JSON Schemas, adapted the current examples, and added a new example that shows the use of uriVariables in the Thing level.

I have also committed td.svg which now looks like this (now there is an arrow between Thing and DataSchema via `uriVariables`):

image

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/pull/1383.html" title="Last updated on Feb 9, 2022, 2:49 PM UTC (a18a9b7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-thing-description/1383/114644a...a18a9b7.html" title="Last updated on Feb 9, 2022, 2:49 PM UTC (a18a9b7)">Diff</a>

### CODE REVIEW

**Review by danielpeintner — APPROVED** [review]
LGTM.

some index.html changes in this PR are there because no `npm render` was called for previous PRs.

File: index.html [hunk 1]
```diff
@@ -1197,7 +1197,11 @@ <h2>Core Vocabulary Definitions</h2>
...
                 name-value pair inside an 
-                <code>AdditionalExpectedResponse</code> object.</td><td>optional</td><td><a>Map</a> of <a href="#dataschema"><code>DataSchema</code></a></td></tr></tbody></table><p><span class="rfc2119-assertion" id=
+                <code>AdditionalExpectedResponse</code> object.</td><td>optional</td><td><a>Map</a> of <a href="#dataschema"><code>DataSchema</code></a></td></tr>
+<tr class="rfc2119-table-assertion" id="td-vocab-uriVariables--Thing"><td><code>uriVariables</code></td><td>Define URI template variables according to [[RFC6570]]  as collection
+                based on DataSchema declarations. The Thing level <code>uriVariables</code> can be used in Thing-level <code>forms</code> or in Interaction Affordances.
+                The individual variables DataSchema cannot be an ObjectSchema or an ArraySchema.
+                If the same variable is also declared in Thing-level <code>uriVariables</code>, Interaction Affordance level variable takes precedence.</td><td>optional</td><td><a>Map</a> of <a href="#dataschema"><code>DataSchema</code></a></td></tr></tbody></table><p><span class="rfc2119-assertion" id=
```
Comment (thjaeckle) [inline index.html:1204]:
```suggestion
                If the same variable is both declared in Thing-level <code>uriVariables</code> and in Interaction Affordance level, the Interaction Affordance level variable takes precedence.</td><td>optional</td><td><a>Map</a> of <a href="#dataschema"><code>DataSchema</code></a></td></tr></tbody></table><p><span class="rfc2119-assertion" id=
```

File: index.html [hunk 2]
```diff
@@ -4377,18 +4383,75 @@ <h3>Top level <code>forms</code></h3>
...
+        }
+        "forms": [{
+            "op": "readallproperties",
+            "href": "https://mything.example.com/properties{?unitSystem}",
+            "contentType": "application/json",
+            "htv:methodName": "GET"
+        }, 
+        {
+            "op": "readmultipleproperties",
+            "href": "https://mylamp.example.com/properties{?fields,unitSystem}",
```
Comment (thjaeckle) [inline index.html:4442]: You defined the uriVariable `propertyNames` above, but referenced `fields` here in the IRI.

File: index.html [hunk 3]
```diff
@@ -4377,18 +4383,75 @@ <h3>Top level <code>forms</code></h3>
...
+            "contentType": "application/json",
+            "htv:methodName": "GET"
+        }]
+    }
+    </pre>
+
+    <p>For a <code>readmultipleproperties</code> operation, an example HTTP GET request to the URI 
+        <code>https://mylamp.example.com/properties?fields=humidity,temperature&unitSystem=metric</code> would return
```
Comment (thjaeckle) [inline index.html:4450]: Same as above - the `fields` uriVariable is not defined.

File: index.template.html [hunk 4]
```diff
@@ -3209,18 +3210,75 @@ <h3>Top level <code>forms</code></h3>
...
+        }
+        "forms": [{
+            "op": "readallproperties",
+            "href": "https://mything.example.com/properties{?unitSystem}",
+            "contentType": "application/json",
+            "htv:methodName": "GET"
+        }, 
+        {
+            "op": "readmultipleproperties",
+            "href": "https://mylamp.example.com/properties{?fields,unitSystem}",
```
Comment (thjaeckle) [inline index.template.html:3269]: Same as above - `fields` uriVariable is not defined above, only `propertyNames`.

File: index.template.html [hunk 5]
```diff
@@ -3209,18 +3210,75 @@ <h3>Top level <code>forms</code></h3>
...
+            "contentType": "application/json",
+            "htv:methodName": "GET"
+        }]
+    }
+    </pre>
+
+    <p>For a <code>readmultipleproperties</code> operation, an example HTTP GET request to the URI 
+        <code>https://mylamp.example.com/properties?fields=humidity,temperature&unitSystem=metric</code> would return
```
Comment (thjaeckle) [inline index.template.html:3277]: Same here - `fields` is not defined.

**Review by thjaeckle — COMMENTED** [review]
Added some inline comments - the example with `fields` as uriVariable does e.g. not match the defined `propertyNames` uriVariable.

**Review by sebastiankb — COMMENTED** [review]
Principle approve this PR. The bug mentioned from @thjaeckle should be fixed

**Review by thjaeckle — APPROVED** [review]

**Review by sebastiankb — APPROVED** [review]

**Comment by sebastiankb** [comment]
from today's TD call decided to merge

**Pull request merged**